lib/workload/stateful/filemanager/filemanager/README.md

```diff
@@ -20,12 +20,45 @@ contain thin wrappers around clients, and any logic should be elsewhere to ensur
 * [env]: Environment variable handling.
 * [error]: Error code associated with filemanager.
 
+Generally code that belongs to a particular cloud service should be put in its own module (e.g. AWS code goes in an `aws` module).
+
+## Event ingestion
+
+The filemanager determines the location of objects in cloud storage by ingesting events like the [AWS S3 events][s3-events].
+However, events for most cloud services only guarantee that events are received at least once. Duplicate events may be received
+or events could be received out of order, and the filemanager should handle this. This needs to occur both in application
+code and at the database level because events can be out of order across Lambda function calls. The aim of the filemanager
+is to maintain a database state that is as correct as possible at the time an event is received.
```

Comment:
Currently the aim is to keep the database as up to date as possible, as soon as an event is received. However, as @brainstorm suggested, it might be better to keep events ordered within a sliding window of time, and wait a certain period (i.e. 5 mins, 1hr, etc) before inserting.

I'm not sure if AWS guarantees anything in terms of events arriving out of order within a certain time period, although in practice I don't think this window would be huge?

Reply:
Worth checking the SQS docs, defaults and guarantees on that.